### ISSUE TITLE
Dataclass takes no arguments in a subinterpreter in 3.12

### ISSUE THREAD

**Issue by peterebden** [python/cpython#110279]
# Bug report

### Bug description:

In 3.12, when I construct an instance of a dataclass in a subinterpreter, I get an "X takes no arguments" error, for cases that work fine in the main interpreter.

A fairly minimal code example:
```c
#include <stdio.h>

#include <Python.h>

const char* code = "\n\
from dataclasses import dataclass\n\
\n\
@dataclass\n\
class MyClass:\n\
  id: str\n\
\n\
c = MyClass(id='abc')\n\
print(c)\n\
";

int main(int argc, char** argv) {
  Py_InitializeEx(0);
  printf("Python version: ");
  fflush(stdout);
  PyRun_SimpleString("import sys; print(sys.version)");
  PyThreadState* main_thread = PyThreadState_Get();

  printf("In main interpreter:\n");
  PyRun_SimpleString(code);

  PyThreadState* interpreter_thread = Py_NewInterpreter();
  printf("\nIn subinterpreter:\n");
  PyRun_SimpleString(code);
  Py_EndInterpreter(interpreter_thread);

  PyThreadState_Swap(main_thread);
  Py_Finalize();
  return 0;
}
```
Compiled with `gcc -o py_dataclass py_dataclass.c -I /usr/include/python3.11 -lpython3.11` I get the expected:
```
Python version: 3.11.5 (main, Aug 25 2023, 13:19:50) [GCC 11.4.0]
In main interpreter:
MyClass(id='abc')

In subinterpreter:
MyClass(id='abc')
```
Compiled with `gcc -o py_dataclass py_dataclass.c -I /usr/include/python3.12 -lpython3.12` I get:
```
Python version: 3.12.0 (main, Oct  2 2023, 15:04:50) [GCC 11.4.0]
In main interpreter:
MyClass(id='abc')

In subinterpreter:
Traceback (most recent call last):
  File "<string>", line 8, in <module>
TypeError: MyClass() takes no arguments
```
This is using the `python3.11-dev` and `python3.12-dev` packages from the deadsnakes PPA.

I'm a little lost about what the interaction between the subinterpreter and the dataclass is that would cause this. The non-mangled Python code is just
```python
from dataclasses import dataclass

@dataclass
class MyClass:
  id: str

c = MyClass(id='abc')
print(c)
```
Let me know if there's any more info I can provide!

### CPython versions tested on:

3.11, 3.12

### Operating systems tested on:

Linux

**Comment by peterebden** [comment]
An interesting datapoint; the following code does work as expected in the subinterpreter:
```python
from dataclasses import dataclass

@dataclass
class MyClass:
  id: str

@dataclass 
class MySubClass(MyClass):
  pass

c = MySubClass(id='abc')
print(c)
```

**Comment by AA-Turner** [comment]
cc: @ericsnowcurrently

**Comment by ericsnowcurrently** [comment]
Any ideas, @ericvsmith?

**Comment by ericsnowcurrently** [comment]
FWIW, it seems to work fine on main.  I was able to reproduce the failure on 3.12:

```python
from test.support import interpreters

interp = interpreters.create()
interp.run("""if True:
    from dataclasses import dataclass

    @dataclass
    class MyClass:
      id: str

    c = MyClass(id='abc')
    print(c)
    """)
```

**Comment by peterebden** [comment]
Ah nice, I didn't know that test module was available to try it out in pure Python.

Also working for me at head, same for our internal usage (which is a bit more complex than the reduced example here - there's something else going wrong there which I'm still digging through).

**Comment by ericvsmith** [comment]
I have no idea what's going on there. I don't have handy access to my dev box right now, but maybe inspecting the signature for `MyClass.__init__` will shed some light.

**Comment by graingert** [comment]
bisected to  ba65a065cf07a7a9f53be61057a090f7311a5ad7 on 3.12

**Comment by ericvsmith** [comment]
Do you know if that made it to the 3.12 branch (so would be in 3.12.1)?

**Comment by graingert** [comment]
I just tried e658413e552b4053014e8d52e3b9153895aebfb3 (the head of 3.12) and I still reproduce the failure

**Comment by graingert** [comment]
as an aside when I was investigating this issue with breakpoint() I got a segfault:

```python
from test.support import interpreters

interp = interpreters.create()
interp.run("""if True:
    from dataclasses import dataclass

    @dataclass
    class MyClass:
      id: str

    breakpoint()
    c = MyClass(id='abc')
    print(c)
    """)
```

python demo.py 
> <string>(9)<module>()
[1]    1338698 segmentation fault (core dumped)  python demo.py

**Comment by anthonyfk** [comment]
For what it's worth, this has broken using Flask in Apache with mod_wsgi, meaning I cannot upgrade to Python 3.12 until this is fixed. So please consider getting this into 3.12.1.

**Comment by ericsnowcurrently** [comment]
I've merged the backport fix.  Could someone verify whether or not this is fixed now.  Thanks!

**Comment by thierryba** [comment]
@ericsnowcurrently if you point me to the patch to try, I can do it. I have the same problems with mod_wsgi in a Django application.

**Comment by ericsnowcurrently** [comment]
The fix is indeed part of 3.12.1.

**Comment by peterebden** [comment]
Thanks, also fixed for me!
